### PR TITLE
Homework_8_Lutsenko_Dmitrii

### PR DESCRIPTION
**Pull request by LutsenkoDm**
# JMH version: 1.25
# VM version: JDK 11.0.13, OpenJDK 64-Bit Server VM, 11.0.13+8-LTS
# VM invoker: C:\Users\lutse\.jdks\corretto-11.0.13\bin\java.exe
# VM options: -Xms1G -Xmx1G
# Warmup: 2 iterations, 10 s each
# Measurement: 3 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: ru.mail.polis.ads.hash.HashTableJmh.benchmarkDefaultHashmap
# Parameters: (TEST_DATA_SIZE = 1000000)

# Run progress: 0,00% complete, ETA 00:05:00
# Warmup Fork: 1 of 1
# Warmup Iteration   1: 1076,328 ms/op
# Warmup Iteration   2: 1198,890 ms/op
Iteration   1: 951,830 ms/op
Iteration   2: 952,911 ms/op
Iteration   3: 1032,348 ms/op

# Run progress: 16,67% complete, ETA 00:08:00
# Fork: 1 of 2
# Warmup Iteration   1: 1212,083 ms/op
# Warmup Iteration   2: 1006,088 ms/op
Iteration   1: 1004,795 ms/op
Iteration   2: 965,211 ms/op
Iteration   3: 889,749 ms/op

# Run progress: 33,33% complete, ETA 00:06:51
# Fork: 2 of 2
# Warmup Iteration   1: 1079,462 ms/op
# Warmup Iteration   2: 998,643 ms/op
Iteration   1: 1218,472 ms/op
Iteration   2: 1080,833 ms/op
Iteration   3: 1179,760 ms/op

Result "ru.mail.polis.ads.hash.HashTableJmh.benchmarkDefaultHashmap":
  1056,470 ±(99.9%) 356,547 ms/op [Average]
  (min, avg, max) = (889,749, 1056,470, 1218,472), stdev = 127,148
  CI (99.9%): [699,923, 1413,016] (assumes normal distribution)

# JMH version: 1.25
# VM version: JDK 11.0.13, OpenJDK 64-Bit Server VM, 11.0.13+8-LTS
# VM invoker: C:\Users\lutse\.jdks\corretto-11.0.13\bin\java.exe
# VM options: -Xms1G -Xmx1G
# Warmup: 2 iterations, 10 s each
# Measurement: 3 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: ru.mail.polis.ads.hash.HashTableJmh.benchmarkHashTableImpl
# Parameters: (TEST_DATA_SIZE = 1000000)

# Run progress: 50,00% complete, ETA 00:05:17
# Warmup Fork: 1 of 1
# Warmup Iteration   1: 3369,451 ms/op
# Warmup Iteration   2: 2834,259 ms/op
Iteration   1: 3344,875 ms/op
Iteration   2: 4330,005 ms/op
Iteration   3: 4504,110 ms/op

# Run progress: 66,67% complete, ETA 00:03:36
# Fork: 1 of 2
# Warmup Iteration   1: 3196,321 ms/op
# Warmup Iteration   2: 3573,468 ms/op
Iteration   1: 3107,497 ms/op
Iteration   2: 3065,134 ms/op
Iteration   3: 2745,255 ms/op

# Run progress: 83,33% complete, ETA 00:01:52
# Fork: 2 of 2
# Warmup Iteration   1: 3452,550 ms/op
# Warmup Iteration   2: 2968,008 ms/op
Iteration   1: 2960,762 ms/op
Iteration   2: 3139,689 ms/op
Iteration   3: 2840,468 ms/op

Result "ru.mail.polis.ads.hash.HashTableJmh.benchmarkHashTableImpl":
  2976,468 ±(99.9%) 441,378 ms/op [Average]
  (min, avg, max) = (2745,255, 2976,468, 3139,689), stdev = 157,400
  CI (99.9%): [2535,090, 3417,846] (assumes normal distribution)

# Run complete. Total time: 00:11:16

REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark                             (TEST_DATA_SIZE)  Mode  Cnt     Score     Error  Units
HashTableJmh.benchmarkDefaultHashmap           1000000  avgt    6  1056,470 ± 356,547  ms/op
HashTableJmh.benchmarkHashTableImpl            1000000  avgt    6  2976,468 ± 441,378  ms/op

Benchmark result is saved to C:\Technopolis\Algorithms\2021-ads\build\results\jmh\results.txt

### CODE REVIEW

**Comment by legohedgehog** [comment]
9 из 10

**Comment by legohedgehog** [comment]
можно исправить комментарии с "-1" в течении суток, если не хочется, можно закрыть ПР

**Comment by LutsenkoDm** [comment]
> можно исправить комментарии с "-1" в течении суток, если не хочется, можно закрыть ПР

Что именно исправлять, вы написали 9 из 10, но за что -1 я не понимаю?

File: src/main/java/ru/mail/polis/ads/hash/HashTableImpl.java [hunk 1]
```diff
@@ -3,33 +3,117 @@
...
     }
 
     @Override
     public boolean isEmpty() {
-        return true;
+        return currentSize == 0;
+    }
+
+    private int hash(Key key) {
+        return Math.abs(key.hashCode() % capacity);
```
[the remark this answers is not in the document]
Comment (LutsenkoDm) [reply src/main/java/ru/mail/polis/ads/hash/HashTableImpl.java]: Исправил

**Comment by LutsenkoDm** [comment]
> можно исправить комментарии с "-1" в течении суток, если не хочется, можно закрыть ПР

Исправил

**Pull request closed without merging**